pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java

```diff
@@ -703,7 +705,23 @@ public boolean isBackwardCompatibleWith(Schema oldSchema) {
       }
       FieldSpec oldSchemaFieldSpec = entry.getValue();
       FieldSpec fieldSpec = getFieldSpecFor(oldSchemaColumnName);
-      if (!fieldSpec.equals(oldSchemaFieldSpec)) {
+
+      if (EqualityUtils.isSameReference(fieldSpec, oldSchemaFieldSpec)) {
+        continue;
+      }
+
+      if ((fieldSpec == null && oldSchemaFieldSpec != null) || (fieldSpec != null && oldSchemaFieldSpec == null)) {
+        return false;
+      }
+
+      boolean isBackward = EqualityUtils.isEqual(fieldSpec.getName(), oldSchemaFieldSpec.getName())
```

Comment:
Let's add a method in `FieldSpec`: `boolean isBackwardCompatibleWith(FieldSpec oldFieldSpec)` so that other sub-classes can override it if necessary

Reply:
Thanks, this is a good point, I addressed it in the second comment.